python/lsst/ap/association/metrics.py

```diff
@@ -0,0 +1,253 @@
+            The total number of unassociated objects.
+        """
+        nNew = 0
+        associated = False
+        for value in values:
+            if value["unassociatedObjects"] is not None:
+                try:
+                    nNew += value["unassociatedObjects"]
+                except TypeError as e:
+                    raise MetricComputationError("Corrupted value of numUnassociatedDiaObjects") from e
+                associated = True
+
+        if associated:
+            return Measurement(self.getOutputMetricName(self.config),
+                               nNew * u.count)
+        else:
+            self.log.info("Nothing to do: no association results found.")
+            return None
+
+    @classmethod
+    def getInputMetadataKeys(cls, config):
+        return {"unassociatedObjects": ".numUnassociatedDiaObjects"}
+
+    @classmethod
+    def getOutputMetricName(cls, config):
+        return "ap_association.numUnassociatedDiaObjects"
+
+
+@register("fracUpdatedDiaObjects")
+class FractionUpdatedDiaObjectsMetricTask(MetadataMetricTask):
+    """Task that computes the fraction of previously created DIAObjects that
+    have a new association in this image, visit, etc..
+    """
+    _DefaultName = "fracUpdatedDiaObjects"
+
+    def makeMeasurement(self, values):
+        """Compute the number of non-updated DIAObjects.
+
+        Parameters
+        ----------
+        values : sequence [`dict` [`str`, `int` or `None`]]
+            A list where each element corresponds to a metadata object passed
+            to `run`. Each `dict` has the following keys:
+
+            ``"updatedObjects"``
+                The number of DIAObjects updated for this image (`int` or
+                `None`). May be `None` if the image was not
+                successfully associated.
+            ``"unassociatedObjects"``
+                The number of DIAObjects not associated with a DiaSource in
+                this image (`int` or `None`). May be `None` if the image was
+                not successfully associated.
+
+        Returns
+        -------
+        measurement : `lsst.verify.Measurement` or `None`
+            The total number of unassociated objects.
+        """
+        nUpdated = 0
+        nUnassociated = 0
+        associated = False
+        for value in values:
+            if value["updatedObjects"] is not None \
+                    and value["unassociatedObjects"] is not None:
+                try:
+                    nUpdated += value["updatedObjects"]
+                    nUnassociated += value["unassociatedObjects"]
+                except TypeError as e:
+                    raise MetricComputationError("Corrupted value of numUpdatedDiaObjects "
+                                                 "or numUnassociatedDiaObjects") from e
+                associated = True
+
+        if associated:
+            if nUpdated <= 0 or nUnassociated <= 0:
+                fraction = 0.0
```

Comment:
This isn't right: if `nUnassociated == 0` and `nUpdated > 0`, `fraction` should be 1.

Reply:
Hmm, indeed. I'll add unit tests for all N = 0 combinations.